### ISSUE TITLE
Providers do not register on macOS 12.6 Monterey

### ISSUE THREAD

**Issue by eggyal** [oxidecomputer/usdt#70]
I've run the `probe-test-build` example as in the repo's README:

<details><summary>console output</summary>
<pre>
% gh repo clone oxidecomputer/usdt
Cloning into 'usdt'...
remote: Enumerating objects: 1859, done.
remote: Counting objects: 100% (386/386), done.
remote: Compressing objects: 100% (161/161), done.
remote: Total 1859 (delta 291), reused 247 (delta 223), pack-reused 1473
Receiving objects: 100% (1859/1859), 412.90 KiB | 2.56 MiB/s, done.
Resolving deltas: 100% (949/949), done.
% cd usdt/probe-test-build
% cargo +nightly run --features asm
   Compiling proc-macro2 v1.0.47
   Compiling unicode-ident v1.0.5
   Compiling quote v1.0.21
   Compiling syn v1.0.105
   Compiling thiserror v1.0.37
   Compiling serde_derive v1.0.148
   Compiling ucd-trie v0.1.5
   Compiling once_cell v1.16.0
   Compiling serde v1.0.148
   Compiling libc v0.2.137
   Compiling version_check v0.9.4
   Compiling serde_json v1.0.89
   Compiling usdt-impl v0.3.4 (/Users/anm/coding/usdt/usdt-impl)
   Compiling itoa v1.0.4
   Compiling ryu v1.0.11
   Compiling byteorder v1.4.3
   Compiling thread-id v4.0.0
   Compiling thiserror-impl v1.0.37
   Compiling pest v2.5.0
   Compiling pest_meta v2.5.0
   Compiling pest_generator v2.5.0
   Compiling pest_derive v2.5.0
   Compiling dtrace-parser v0.1.14 (/Users/anm/coding/usdt/dtrace-parser)
   Compiling serde_tokenstream v0.1.5
   Compiling usdt-attr-macro v0.3.4 (/Users/anm/coding/usdt/usdt-attr-macro)
   Compiling usdt-macro v0.3.4 (/Users/anm/coding/usdt/usdt-macro)
   Compiling usdt v0.3.4 (/Users/anm/coding/usdt/usdt)
   Compiling probe-test-build v0.1.0 (/Users/anm/coding/usdt/probe-test-build)
    Finished dev [unoptimized + debuginfo] target(s) in 36.50s
     Running `/Users/anm/coding/usdt/target/debug/probe-test-build
</pre>
</details>

But the providers do not appear to be registered with dtrace:

```zsh
% sudo dtrace -l -n 'my_provider*:::'
   ID   PROVIDER            MODULE                          FUNCTION NAME
dtrace: failed to match my_provider*:::: No probe matches description
```

I have tried SIP enabled both with and without dtrace.  For the avoidance of doubt, had this issue with my own providers before I attempted to verify with the example too.

Does one need to do anything further to register/use USDT providers on Monterey?

**Comment by rzezeski** [comment]
Is the process still running when you run `dtrace -l`?

You might have better luck by running the following dtrace first, and then run your program while dtrace is still executing.

```
dtrace -Z -n 'my_provider*:::'
```

**Comment by eggyal** [comment]
Unfortunately, no progress yet:

<details><summary>Results of above suggestions</summary>

> Is the process still running when you run `dtrace -l`?

Yes, it was.

> You might have better luck by running the following dtrace first, and then run your program while dtrace is still executing.
>
> ```
> dtrace -Z -n 'my_provider*:::'
> ```

Output below (`probe-test-build` was run after "matched 0 probes" and killed some time later, before the ^C):

```zsh
% sudo dtrace -Z -n 'my_provider*:::'
dtrace: description 'my_provider*:::' matched 0 probes
^C
```

> Another option would be to run the test directly with DTrace:
> 
> ```sh
> $ sudo dtrace -n 'my_provider*:::' -c './target/debug/probe-test-build'
> ```

Output below:

```zsh
% sudo dtrace -n 'my_provider*:::' -c './target/debug/probe-test-build'
dtrace: invalid probe specifier my_provider*:::: probe description my_provider*::: does not match any probes
```
</details>

I checked the build script output in `target/debug/build/usdt-impl-2f0bd06f84bbd7bb/output`:
<pre>
cargo:rerun-if-changed=build.rs
cargo:rustc-cfg=usdt_need_feat_asm_sym
cargo:rustc-cfg=usdt_stable_asm
cargo:rustc-cfg=usdt_backend_linker
</pre>

Which looks reasonable to me?

Then I ran `cargo expand` on `probe-test-build`, which gave:

<details><summary>Expanded crate</summary>
<pre>
#![feature(prelude_import)]
//! An example using the `usdt` crate, generating the probes via a build script.
#![feature(asm_sym)]
#[prelude_import]
use std::prelude::rust_2018::*;
#[macro_use]
extern crate std;
use std::thread::sleep;
use std::time::Duration;
use usdt::register_probes;
pub(crate) mod test {
    #[allow(unused_imports)]
    pub(crate) use start_work;
    #[allow(unused_imports)]
    pub(crate) use stop_work;
}
fn main() {
    let duration = Duration::from_secs(1);
    let mut counter: u8 = 0;
    register_probes().unwrap();
    loop {
        {
            let __usdt_private_args_lambda = || (counter);
            #[allow(unused_imports)]
            #[allow(non_snake_case)]
            fn __usdt_private_test_start_work_type_check(_: impl ::std::borrow::Borrow<u8>) {}
            let _ = || {
                let args = (__usdt_private_args_lambda.clone()(),);
                __usdt_private_test_start_work_type_check(args.0);
            };
            extern "C" {
                #[allow(unused)]
                #[link_name = "__dtrace_stability$test$v1$1_1_0_1_1_0_1_1_0_1_1_0_1_1_0"]
                fn stability();
                #[allow(unused)]
                #[link_name = "__dtrace_typedefs$test$v2"]
                fn typedefs();
                #[allow(unused)]
                #[link_name = "__dtrace_isenabled$test$start_work$v1"]
                fn test_start_work_enabled() -> i32;
                #[allow(unused)]
                #[link_name = "__dtrace_probe$test$start_work$v1$75696e74385f74"]
                fn __start_work(_: ::std::os::raw::c_uchar);
            }
            unsafe {
                if test_start_work_enabled() != 0 {
                    let args = (__usdt_private_args_lambda(),);
                    let arg_0 = (*<_ as ::std::borrow::Borrow<u8>>::borrow(&args.0) as i64);
                    asm ! (".reference {0}\ncall {1}\n.reference {2}" , sym typedefs , sym __start_work , sym stability , in ("rdi") (arg_0) , options (nomem , preserves_flags , nostack));
                }
            }
        };
        sleep(duration);
        {
            let __usdt_private_args_lambda = || {
                (
                    {
                        let res = ::alloc::fmt::format(::core::fmt::Arguments::new_v1(
                            &["the probe has fired "],
                            &[::core::fmt::ArgumentV1::new_display(&counter)],
                        ));
                        res
                    },
                    counter,
                    {
                        let res = ::alloc::fmt::format(::core::fmt::Arguments::new_v1(
                            &[""],
                            &[::core::fmt::ArgumentV1::new_lower_hex(&counter)],
                        ));
                        res
                    },
                )
            };
            #[allow(unused_imports)]
            #[allow(non_snake_case)]
            fn __usdt_private_test_stop_work_type_check(
                _: impl AsRef<str>,
                _: impl ::std::borrow::Borrow<u8>,
                _: impl AsRef<str>,
            ) {
            }
            let _ = || {
                let args = __usdt_private_args_lambda.clone()();
                __usdt_private_test_stop_work_type_check(args.0, args.1, args.2);
            };
            extern "C" {
                #[allow(unused)]
                #[link_name = "__dtrace_stability$test$v1$1_1_0_1_1_0_1_1_0_1_1_0_1_1_0"]
                fn stability();
                #[allow(unused)]
                #[link_name = "__dtrace_typedefs$test$v2"]
                fn typedefs();
                #[allow(unused)]
                #[link_name = "__dtrace_isenabled$test$stop_work$v1"]
                fn test_stop_work_enabled() -> i32;
                #[allow(unused)]
                #[link_name = "__dtrace_probe$test$stop_work$v1$63686172202a$75696e74385f74$63686172202a"]
                fn __stop_work(
                    _: *const ::std::os::raw::c_char,
                    _: ::std::os::raw::c_uchar,
                    _: *const ::std::os::raw::c_char,
                );
            }
            unsafe {
                if test_stop_work_enabled() != 0 {
                    let args = __usdt_private_args_lambda();
                    let arg_0 = [(args.0.as_ref() as &str).as_bytes(), &[0_u8]].concat();
                    let arg_1 = (*<_ as ::std::borrow::Borrow<u8>>::borrow(&args.1) as i64);
                    let arg_2 = [(args.2.as_ref() as &str).as_bytes(), &[0_u8]].concat();
                    asm ! (".reference {0}\ncall {1}\n.reference {2}" , sym typedefs , sym __stop_work , sym stability , in ("rdi") (arg_0 . as_ptr () as i64) , in ("rsi") (arg_1) , in ("rdx") (arg_2 . as_ptr () as i64) , options (nomem , preserves_flags , nostack));
                }
            }
        };
        counter = counter.wrapping_add(1);
    }
}
</pre>
</details>

I can't say whether this is reasonable or not, but certainly nothing jumps out as an obvious failure in constructing the to-be-compiled source.

Grateful for any further thoughts or suggestions!

**Comment by bnaecker** [comment]
Welp, think I found the problem! The provider isn't name `my_provider` it's `test`. I'm not sure where that came from, but I just ran with what you'd written. I believe there are _other_ examples or tests that use that, but the `probe-test-build` crate does not. On my (older) macOS system:

```bash
bnaecker@sandstone : ~/file-cabinet/oxide/usdt $ sudo dtrace -n 'test*:::' -c ./target/debug/probe-test-build
Password:
dtrace: system integrity protection is on, some features will not be available

dtrace: description 'test*:::' matched 2 probes
CPU     ID                    FUNCTION:NAME
  2  35486 _ZN16probe_test_build4main17hcb8b657ee6d5a9eaE:start_work
  2  35487 _ZN16probe_test_build4main17hcb8b657ee6d5a9eaE:stop_work
  2  35486 _ZN16probe_test_build4main17hcb8b657ee6d5a9eaE:start_work
  0  35487 _ZN16probe_test_build4main17hcb8b657ee6d5a9eaE:stop_work
  0  35486 _ZN16probe_test_build4main17hcb8b657ee6d5a9eaE:start_work
^C
  2  35487 _ZN16probe_test_build4main17hcb8b657ee6d5a9eaE:stop_work
  2  35486 _ZN16probe_test_build4main17hcb8b657ee6d5a9eaE:start_work

```

Does that work for you?

**Comment by bnaecker** [comment]
Also as an aside, the `dusty` binary that comes with the crate is a quick way to see the probes an object file exports. That shows the below on that binary:

```bash
Section {
    ident: Ident {
        magic: [
            127,
            68,
            79,
            70,
        ],
        model: LP64,
        encoding: LittleEndian,
        version: 3,
        dif_vers: 2,
        dif_ireg: 8,
        dif_treg: 8,
    },
    providers: {
        "test": Provider {
            name: "test",
            probes: {
                "start_work": Probe {
                    name: "start_work",
                    function: "_ZN16probe_test_build4main17hcb8b657ee6d5a9eaE",
                    address: 0,
                    offsets: [
                        4294664454,
                    ],
                    enabled_offsets: [
                        4294664347,
                    ],
                    arguments: [
                        "uint8_t",
                    ],
                },
                "stop_work": Probe {
                    name: "stop_work",
                    function: "_ZN16probe_test_build4main17hcb8b657ee6d5a9eaE",
                    address: 0,
                    offsets: [
                        4294665069,
                    ],
                    enabled_offsets: [
                        4294664475,
                    ],
                    arguments: [
                        "char *",
                        "uint8_t",
                        "char *",
                    ],
                },
            },
        },
    },
}
```

**Comment by eggyal** [comment]
In the interim, I noticed I was still on a rather old version of nightly.  I've since upgraded, and now (using the correct provider name!) it is indeed working.  That said, I'm still having problems with my own crate, but dusty (thanks for that pointer!) shows that my binary doesn't contain any probe information so I'll dig deeper into that.

Thank you for all the time and input—apologies that it was such a minor oversight one war or the other!  Closing.
